Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,379 @@
+-/
+def set (v : Vector α n) (i : Fin n) (x : α) : Vector α n :=
+  ⟨v.toArray.set (v.size_eq.symm ▸ i) x, proof⟩
+  where
+    proof := by {rw [size_set]; exact v.size_eq}
+
+/--
+Set an element in a vector, or do nothing if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def setD (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.setD i x, proof⟩
+  where
+    proof := by
+      rw [size_setD]
+      exact v.size_eq
+
+/--
+Set an element in an array, or panic if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def set! (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.set! i x, proof⟩
+  where
+    proof := by
+      rw [set!_is_setD, size_setD]
+      exact v.size_eq
+
+/-- Appending a vector to another. -/
+def append {n m : Nat} : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+instance : HAppend (Vector α n) (Vector α m) (Vector α (n + m)) where
+  hAppend := append
+
+/-- `Vector.head` produces the head of a vector -/
+abbrev head (v : Vector α (n+1)) :=
+  v.toArray.get ⟨0, v.size_eq.symm ▸ Nat.zero_lt_succ n⟩
+
+/--
+`Vector.cons` is the list cons pattern for vectors.
+
+**Warning** : This is an O(n) operation that prepends an element to a vector. It needs to shift all its previous elements one index to the right.
+-/
+def cons (x : α) (xs : Vector α n) : Vector α (n + 1) := Nat.add_comm 1 n ▸ (#v[x] ++ xs)
+
+/-- Make a vector of size `n` with all cells containing `v` -/
+def mkVector (n : Nat) (v : α) : Vector α n :=
+  ⟨mkArray n v, proof⟩
+  where
+    proof := by
+      rw [size_mkArray]
+
+/--
+`extract v start halt` Returns the slice of `v` from indices `start` to `stop` (exclusive). If `start` is greater or equal to `stop`, the result is empty. If `stop` is greater than the size of `v`, the size is used instead.
+-/
+def extract (v : Vector α n)
+  (start : Nat) (halt : Nat) : Vector α (min halt n - start) :=
+  ⟨Array.extract v.toArray start halt, proof⟩
+  where
+    proof := by
+      rw [size_extract, v.size_eq]
+
+/-- Custom eliminator for `Vector α n` through Arrays -/
+@[elab_as_elim]
+def elimAsArray {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : Array α, motive ⟨a, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨a, rfl⟩ => mk a
+
+/-- Custom eliminator for `Vector α n` throughLists -/
+@[elab_as_elim]
+def elimAsList {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : List α, motive ⟨⟨a⟩, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨⟨a⟩, rfl⟩ => mk a
+
+/-- Map a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨a, h⟩ => ⟨Array.map f a, by simp [*]⟩
+
+/-- Mapping two vectors under a curried function of two variables. -/
+def zipWith (f : α → β → φ) : Vector α n → Vector β n → Vector φ n
```

Comment:
Can you change the argument order to match `Array.zipWith`? (i.e. `f` comes last.)

Reply:
Done.